orm/src/main/java/org/hibernate/search/query/hibernate/impl/CriteriaObjectInitializer.java

```diff
@@ -105,33 +111,124 @@ private void setCriteriaTimeout(Criteria criteria, TimeoutManager timeoutManager
+
+			for ( Entry<Class<?>, List<EntityInfo>> infosOfIdSpace : infosByIdSpace.entrySet() ) {
+				Criteria criteria = objectInitializationContext.getSession().createCriteria( infosOfIdSpace.getKey() );
+				criteria.add( getIdListCriterion( infosOfIdSpace.getValue(), objectInitializationContext ) );
+				criterias.add( criteria );
+			}
+
+			return criterias;
+		}
+	}
+
+	/**
+	 * Returns a {@link Criterion} for fetching all the given entity infos. If needed, this criterion will contain a
+	 * {@link Disjunction} for fetching the infos in chunks of {@link CriteriaObjectInitializer#MAX_IN_CLAUSE} elements.
+	 */
+	private Criterion getIdListCriterion(List<EntityInfo> entityInfos, ObjectInitializationContext objectInitializationContext) {
 		DocumentBuilderIndexedEntity documentBuilder = getDocumentBuilder(
-				objectInitializationContext.getEntityType(),
+				entityInfos.iterator().next().getClazz(),
 				objectInitializationContext.getExtendedSearchintegrator()
 		);
 		String idName = documentBuilder.getIdentifierName();
 		Disjunction disjunction = Restrictions.disjunction();
 
+		int maxResults = entityInfos.size();
 		int loop = maxResults / MAX_IN_CLAUSE;
 		boolean exact = maxResults % MAX_IN_CLAUSE == 0;
 		if ( !exact ) {
 			loop++;
 		}
 		for ( int index = 0; index < loop; index++ ) {
-			int max = index * MAX_IN_CLAUSE + MAX_IN_CLAUSE <= maxResults ?
-					index * MAX_IN_CLAUSE + MAX_IN_CLAUSE :
-					maxResults;
+			int max = Math.min( index * MAX_IN_CLAUSE + MAX_IN_CLAUSE, maxResults );
+
 			List<Serializable> ids = new ArrayList<>( max - index * MAX_IN_CLAUSE );
 			for ( int entityInfoIndex = index * MAX_IN_CLAUSE; entityInfoIndex < max; entityInfoIndex++ ) {
-				ids.add( entityInfos[entityInfoIndex].getId() );
+				ids.add( entityInfos.get( entityInfoIndex ).getId() );
 			}
 			disjunction.add( Restrictions.in( idName, ids ) );
 		}
-		criteria.add( disjunction );
+
+		return disjunction;
+	}
+
+	/**
+	 * Groups the given entity infos by id spaces. An id space is a set of entity types which share the same id
+	 * property, e.g. defined in a common super-entity.
+	 *
+	 * @return The given entity infos, keyed by the root entity type of id spaces
+	 */
+	private Map<Class<?>, List<EntityInfo>> groupInfosByIdSpace(EntityInfo[] entityInfos, ObjectInitializationContext objectInitializationContext) {
+		ServiceManager serviceManager = objectInitializationContext.getExtendedSearchintegrator().getServiceManager();
+		IdUniquenessResolver resolver = serviceManager.requestService( IdUniquenessResolver.class );
+		SessionFactoryImplementor sessionFactory = (SessionFactoryImplementor) objectInitializationContext.getSession().getSessionFactory();
+
+		try {
+			Map<Class<?>, List<EntityInfo>> infosByIdSpace = new HashMap<>();
+
+			for ( EntityInfo entityInfo : entityInfos ) {
+				boolean addedToExistingGroup = processEntityInfo( infosByIdSpace, resolver, entityInfo );
+
+				if ( !addedToExistingGroup ) {
```

Comment:
why do you need to return a boolean from `processEntityInfo`? Could you not handle the adding of the new `EntityInfo` list in `processEntityInfo` directly?